pkg/tempo/promsdprocessor/prom_sd_processor.go

```diff
@@ -0,0 +1,220 @@
+
+	logger log.Logger
+}
+
+func newTraceProcessor(nextConsumer consumer.TraceConsumer, scrapeConfigs []*config.ScrapeConfig) (component.TraceProcessor, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	logger := log.With(util.Logger, "component", "tempo service disco")
+	mgr := discovery.NewManager(ctx, logger, discovery.Name("tempo service disco"))
+
+	relabelConfigs := map[string][]*relabel.Config{}
+	cfg := map[string]sd_config.ServiceDiscoveryConfig{}
+	for _, v := range scrapeConfigs {
+		cfg[v.JobName] = v.ServiceDiscoveryConfig
+		relabelConfigs[v.JobName] = v.RelabelConfigs
+	}
+
+	err := mgr.ApplyConfig(cfg)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if nextConsumer == nil {
+		cancel()
+		return nil, componenterror.ErrNilNextConsumer
+	}
+	return &promServiceDiscoProcessor{
+		nextConsumer:     nextConsumer,
+		discoveryMgr:     mgr,
+		discoveryMgrStop: cancel,
+		relabelConfigs:   relabelConfigs,
+		hostLabels:       make(map[string]model.LabelSet),
+		logger:           util.Logger,
+	}, nil
+}
+
+func (p *promServiceDiscoProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		if rs.IsNil() {
+			continue
+		}
+		r := rs.Resource()
+		if r.IsNil() {
+			continue
+		}
+
+		p.processAttributes(r.Attributes())
+	}
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *promServiceDiscoProcessor) processAttributes(attrs pdata.AttributeMap) {
+	// find the ip
+	ipTagNames := []string{
+		"ip",          // jaeger/opentracing? default
+		"net.host.ip", // otel semantics for host ip
+	}
+
+	var ip string
+	for _, name := range ipTagNames {
+		val, ok := attrs.Get(name)
+		if !ok {
+			continue
+		}
+
+		ip = val.StringVal()
+		break
+	}
+
+	// have to have an ip for labels lookup
+	if ip == "" {
+		return
+	}
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	labels, ok := p.hostLabels[ip]
+	if !ok {
+		level.Debug(p.logger).Log("msg", "unable to find matching hostLabels", "ip", ip)
+		return
+	}
+
+	for k, v := range labels {
+		attrs.UpsertString(string(k), string(v))
+	}
+}
+
+func (p *promServiceDiscoProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (p *promServiceDiscoProcessor) Start(_ context.Context, _ component.Host) error {
+	go p.watchServiceDiscovery()
+
+	go func() {
+		err := p.discoveryMgr.Run()
+		if err != nil {
+			level.Error(p.logger).Log("msg", "failed to start prom svc disco.  relabeling disabled", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (p *promServiceDiscoProcessor) Shutdown(context.Context) error {
+	if p.discoveryMgrStop != nil {
+		p.discoveryMgrStop()
+	}
+	return nil
+}
+
+func (p *promServiceDiscoProcessor) watchServiceDiscovery() {
+	for targetGoups := range p.discoveryMgr.SyncCh() {
```

Comment:
I think this will need a `select`; the SyncCh doesn't close when the context is canceled, so you'll need to break out of this goroutine when the processor is shut down. 

(You'll also need to close `SyncCh` yourself, Prometheus never closes it)

Reply:
(Nevermind, you don't need to close the channel yourself, the go GC will handle that for you)